ext/pip.ts

```diff
@@ -0,0 +1,170 @@
+import { kw, python, PythonError } from "../mod.ts";
+
+import { join } from "https://deno.land/std@0.197.0/path/mod.ts";
+import { ensureDir } from "https://deno.land/std@0.197.0/fs/mod.ts";
+import { green, yellow } from "https://deno.land/std@0.197.0/fmt/colors.ts";
+
+import type { CacheLocation } from "https://deno.land/x/plug@1.0.2/types.ts";
+import { ensureCacheLocation } from "https://deno.land/x/plug@1.0.2/download.ts";
+import { hash } from "https://deno.land/x/plug@1.0.2/util.ts";
+
+const sys = python.import("sys");
+const runpy = python.import("runpy");
+const importlib = python.import("importlib");
+
+// https://packaging.python.org/en/latest/specifications/name-normalization/
+const MODULE_REGEX =
+  /^([a-z0-9]|[a-z0-9][a-z0-9._-]*[a-z0-9])([^a-z0-9._-].*)?$/i;
+
+function normalizeModuleName(name: string) {
+  return name.replaceAll(/[-_.]+/g, "-").toLowerCase();
+}
+
+function getModuleNameAndVersion(module: string): {
+  name: string;
+  version?: string;
+} {
+  const match = module.match(MODULE_REGEX);
+  const name = match?.[1];
+  const version = match?.[2];
+
+  if (name == null) {
+    throw new TypeError("Could not match any valid pip module name");
+  }
+
+  return {
+    name: normalizeModuleName(name),
+    version,
+  };
+}
+
+export class Pip {
+  #cacheLocation: Promise<string>;
+
+  constructor(location: CacheLocation) {
+    this.#cacheLocation = Promise.all([
+      ensureCacheLocation(location),
+      globalThis.location !== undefined
+        ? hash(globalThis.location.href)
+        : Promise.resolve("pip"),
+    ]).then(async (parts) => {
+      const cacheLocation = join(...parts);
+      await ensureDir(cacheLocation);
+
+      if (!(cacheLocation in sys.path)) {
+        sys.path.insert(0, cacheLocation);
+      }
+
+      return cacheLocation;
+    });
+  }
+
+  /**
+   * Install a Python module using the `pip` package manager.
+   *
+   * @param module The Python module which you wish to install
+   *
+   * @example
+   * ```ts
+   * import { python } from "https://deno.land/x/python/mod.ts";
+   * import { install } from "https://deno.land/x/python/ext/pip.ts";
+   *
+   * await install("fuckit");
+   * const fuckit = python.import("fuckit");
```

Comment:
🤨

Reply:
Oops, had a really frustrating evening to say the least.